## QTM 385

***

## Homework 3

Student: Mina Lee [2298297]

In this homework, we will do some very elementary text analysis. The idea is to:

1. Open a Gutenberg Project Book
2. Count how many lines the Book has.
3. Count how many words the Book has.

The [Project Gutenberg](https://www.gutenberg.org) stores thousands of texts that are free for access, download, and read in your own device. In this homework, we will read the book "Pride and Prejudice," by Jane Austen. In case you want to know more about the book, here is a [Wikipedia entry](https://en.wikipedia.org/wiki/Pride_and_Prejudice) describing it.


***

### Question 1

We start by opening the book. The `txt` version of the book is in the following link: https://www.gutenberg.org/files/1342/1342-0.txt

To read the text from this URL we will use the function `urlopen` in the module `urllib.request`. Read this [Stackoverflow entry](https://stackoverflow.com/questions/47842388/read-text-files-from-website-with-python), and adapt the code in the first answer to _open the book_, _display the first line_, and then _`break`_.

In [1]:
# Your answer here
import urllib.request

In [13]:
book = urllib.request.urlopen('https://www.gutenberg.org/files/1342/1342-0.txt')

In [14]:
i = 0
for line in book:
    print(line)
    i += 1
    if i == 1:
        break

b'\xef\xbb\xbfThe Project Gutenberg eBook of Pride and Prejudice, by Jane Austen\r\n'



***

### Question 2

Note that at the beginning and the end of the first line, we have some strange characters (e. g. `\xef\xbb\xbf`). This happens because of the [character encoding](https://en.wikipedia.org/wiki/Character_encoding). To fix this, we need to use `decode('utf-8')`. For instance, check out the following code:

In [30]:
# Here is my first line
my_line = b'\xef\xbb\xbfThe Project Gutenberg eBook of Pride and Prejudice, by Jane Austen\r\n'

# My line before decoding
print(my_line, end = '\n\n')

# My line after decoding
print(my_line.decode('utf-8'))

b'\xef\xbb\xbfThe Project Gutenberg eBook of Pride and Prejudice, by Jane Austen\r\n'

The Project Gutenberg eBook of Pride and Prejudice, by Jane Austen



To fix this problem, we will create a list with all the lines in the book, decoded to use the UTF-8 encoding. Your job, step by step, is to:

1. Reopen the book
2. Create an empty list
3. Use an for-loop to go over each line of the text
4. Inside the for-loop, change the encoding of line, and store in the list (hint: you could use `append()` to do this)

In the end, display the first line.

In [39]:
# Your answer here
import urllib.request 
book = urllib.request.urlopen('https://www.gutenberg.org/files/1342/1342-0.txt')
list=[]
for line in book:
    list.append(line.decode('utf-8'))
    linenumber = 1
    i = 1
    if i == linenumber:
        print(line.decode('utf-8'))
        break

The Project Gutenberg eBook of Pride and Prejudice, by Jane Austen




***

### Question 3

How many lines the book has?

In [10]:
# Your answer here
import urllib.request 
book= urllib.request.urlopen('https://www.gutenberg.org/files/1342/1342-0.txt')
count = 0
for line in book:
    count += 1
    
count

14580

There are 14,580 lines in the book.


***

### Question 4

Now, to count how many words there is in the book, we need a function that splits a text line into its words. For example, to split the first line:

In [5]:
# New decoded line:
my_first_line = "This is my first line"

# My splitted line
print(my_first_line.split())

['This', 'is', 'my', 'first', 'line']


Split the 385th line of the text, and show how many words it has (hint: python starts counting at zero): 

In [9]:
# Your answer here
import urllib.request 
book= urllib.request.urlopen('https://www.gutenberg.org/files/1342/1342-0.txt')
y=0
x=0
for line in book:
    z=line.decode('utf-8')
    y += 1
    if y==385:
        z.split()
        x += len(z.split())
        
x

13

The 385th line of the text has 13 words.


***

### Question 5

Now, count how many words the book has.

In [18]:
# Your answer here
import urllib.request 
book= urllib.request.urlopen('https://www.gutenberg.org/files/1342/1342-0.txt')
count = 0
for line in book:
    count += len(line.split())
    
count

124743

There are 124,743 words in the book.

**That's all, folks!**